src/Raven.Server/Documents/ResourceCache.cs

```diff
@@ -86,44 +101,58 @@ private bool UnlikelyTryGet(StringSegment resourceName, out Task<TResource> reso
 
         public bool TryRemove(StringSegment resourceName, Task<TResource> resourceTask)
         {
-            if (_caseInsensitive.TryRemove(new KeyValuePair<StringSegment, Task<TResource>>(resourceName, resourceTask)) == false)
-                return false;
-
-            _resourceDetails.TryRemove(resourceTask, out _);
-
             lock (this)
             {
-                RemoveCaseSensitive(resourceName, resourceTask);
+                if (TryRemoveInternal(resourceName, resourceTask) == false) 
+                    return false;
+
+                // We need to refresh all of them.
+                _readonlyCaseSensitive = _caseSensitive.ToFrozenDictionary();
+                _readonlyCaseInsensitive = _caseInsensitive.ToFrozenDictionary();
+                _readonlyResourceDetails = _resourceDetails.ToFrozenDictionary();
             }
 
             return true;
         }
 
-        public bool TryGetAndRemove(StringSegment resourceName, out Task<TResource> resourceTask)
+        private bool TryRemoveInternal(StringSegment resourceName, Task<TResource> resourceTask)
         {
-            if (TryGetValue(resourceName, out resourceTask) == false)
+            if (_caseInsensitive.Remove(resourceName, out _) == false)
                 return false;
 
-            if (TryRemove(resourceName, resourceTask) == false)
-                return false;
+            _resourceDetails.Remove(resourceTask, out _);
 
+            RemoveCaseSensitive(resourceName);
             return true;
         }
 
-        private void RemoveCaseSensitive(StringSegment resourceName, Task<TResource> resourceTask)
+
+        private void RemoveCaseSensitive(StringSegment resourceName)
         {
             if (_mappings.TryGetValue(resourceName, out ConcurrentSet<StringSegment> mappings))
             {
                 foreach (var mapping in mappings)
                 {
-                    _caseSensitive.TryRemove(new KeyValuePair<StringSegment, Task<TResource>>(mapping, resourceTask));
+                    _caseSensitive.Remove(mapping, out _);
```

Comment:
Pretty sure that this is a problem - note that we were previously checking that both the key & value matched.
IIRC, that was a bug fix, see:

https://github.com/ravendb/ravendb/commit/34c9fcb5a111b352795f30ce24bd91d7a68bfe31